cmd/osm-controller/osm-controller.go

```diff
@@ -218,23 +224,56 @@ func main() {
 	// TODO(draychev): figure out the NS and POD
 	metricsStore := metricsstore.NewMetricStore("TBD_NameSpace", "TBD_PodName")
 
-	// Expose /debug endpoints and data only if the enableDebugServer flag is enabled
-	var debugServer debugger.DebugServer
-	if cfg.IsDebugServerEnabled() {
-		debugServer = debugger.NewDebugServer(certDebugger, xdsServer, meshCatalog, kubeConfig, kubeClient, cfg, kubernetesClient)
-	}
-
 	funcProbes := []health.Probes{xdsServer}
 	httpProbes := getHTTPHealthProbes()
-	httpServer := httpserver.NewHTTPServer(funcProbes, httpProbes, metricsStore, constants.MetricsServerPort, debugServer)
+	httpServer := httpserver.NewHTTPServer(funcProbes, httpProbes, metricsStore, constants.MetricsServerPort)
 	httpServer.Start()
 
-	// Wait for exit handler signal
-	<-stop
+	// Expose /debug endpoints and data only if the enableDebugServer flag is enabled
+	debugConfig := debugger.NewDebugConfig(certDebugger, xdsServer, meshCatalog, kubeConfig, kubeClient, cfg, kubernetesClient)
+	c := controller{
+		debugServerRunning: !cfg.IsDebugServerEnabled(),
+		debugComponents:    debugConfig,
+		debugServer:        httpserver.NewDebugHTTPServer(debugConfig, constants.DebugPort),
+	}
+
+	c.initDebugServer(cfg, stop)
 
 	log.Info().Msg("Goodbye!")
 }
 
+func (c *controller) initDebugServer(cfg configurator.Configurator, stop chan struct{}) {
+	errCh := make(chan error)
+	go c.configureDebugServer(cfg, errCh)
+
+	select {
+	case err := <-errCh:
+		log.Error().Err(err).Msg("Unable to configure debug server")
+
+		// Wait for exit handler signal
+	case <-stop:
```

Comment:
I think we could remove watching for signals on via `<-stop` in this func and move it back up to the main func

Reply:
Yes please move this to main, else this function could read the channel first and main will never exit when it should.